incontext.py

```diff
@@ -50,11 +88,33 @@ def context_function(name=None):
     If `name` is not specified, the function name is used.
     """
     def decorator(f):
-        CONTEXT_FUNCTIONS[name if name is not None else f.__name__] = f
+        _PLUGINS.add_plugin(PLUGIN_TYPE_CONTEXT_FUNCTION, name if name is not None else f.__name__, f)
         return f
     return decorator
 
 
+def command(name, help=None, arguments=[]):
+    """
+    Register a new command.
+    """
+    def decorator(f):
+        def add_command_callback(incontext, parser):
+            for argument in arguments:
+                parser.add_argument(**argument)
+            def callback_wrapper(options):
+                return f(incontext, options)
+            return callback_wrapper
+        _PLUGINS.add_plugin(PLUGIN_TYPE_COMMAND, name, wrap_add_command(name, add_command_callback, help))
+        return f
+    return decorator
+    
+    
+def wrap_add_command(name, f, help):
+    def add_command(incontext):
+        return incontext._add_command(name, f, help)
+    return add_command
```

Comment:
It would be cleaner to create a new object which captures the properties associated with the argument and adds these to the plugin; they will be easier to enumerate (adding in a description, etc) for a command which lists the available plugins, and are less magic than a closure which has the effect of hiding away much of the details about what's going on, especially at the call site.